### PR TITLE
Moved Request Validation and Token check into Middleware, added Route to get Players by Alliance 

### PR DESCRIPTION
**Pull request by kejith**
Instead of doing this for every Route
```js
  const token = req.headers.token
  if (!await Token.isValid(token)) {
    return res.status(401).json({ message: 'Provided token is invalid.' })
  }
  const validationErrors = validationResult(req)
  if (!validationErrors.isEmpty()) {
    return res.status(400).json({ errors: validationErrors.array() })
  }
```
We just import `const { checkToken, validateRequest } = require('../middleware')`. Then incorporate them into the function declaration as arguments as seen above.
```js
router.post('/v1/sample', ... , validateRequest, checkToken, async (req, res) => {
// ...
}
```

I also created a new route called `v1/players/alliance/Alliance1,Alliance2` that allows you to obtain all of the members of a certain Alliance, or even multiple Alliances.

Due to the fact that we accessed the DB n times for n usernames, I also decreased function calls to the DB and on "Players.getByName()".

Please feel free to suggest changes.

### CODE REVIEW

File: lib/data/player.js [hunk 1]
```diff
@@ -81,13 +81,49 @@ module.exports = class Player {
...
-    const p = Player.fromDB(response[0])
-    return p
+    let response, players
+    if (Array.isArray(playerName)) {
+      response = await knex('players')
+        .select('*')
+        .whereIn('player_name', playerName)
+        .orderBy('id', 'desc')
+
+      players = []
```
Comment (anoff) [inline lib/data/player.js:91]: This is awesome, but thinking about type safety the function should have a consistent API, not taking an array or a single name.

Two ways I can think of..maybe there are more:
1. use `getByNames (playerNames)` and it always is an array
2. have two functions, `getByName` takes a single string, `getByNames` takes an array of strings

File: lib/data/player.js [hunk 2]
```diff
@@ -81,13 +81,49 @@ module.exports = class Player {
...
+
+      players = Player.fromDB(response[0])
+    }
+
+    return players
+  }
+
+  static async getByAllianceName (allianceName) {
+    let response
+    if (Array.isArray(allianceName)) {
```
Comment (anoff) [inline lib/data/player.js:110]: same as above, type should be fixed. maybe also add a jsdoc to clarify it ;-)

File: lib/middleware.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const Token = require('./data/token')
+const { validationResult } = require('express-validator')
+
+const checkToken = async (req, res, next) => {
+  const token = req.headers.token
+  if (!await Token.isValid(token)) {
+    return res.status(401).json({ message: 'Provided token is invalid.' })
+  }
+  next()
+}
```
Comment (anoff) [inline lib/middleware.js:10]: ❤️ great idea

**Review by anoff — CHANGES_REQUESTED** [review]

**Comment by anoff** [comment]
Do you have plans to add a frontend feature making use of the alliance or is it just to dump the data?

File: lib/data/player.js (hunk 1 above)
Comment (kejith) [reply lib/data/player.js:91]: This sound good, will split it into two functions.

**Comment by kejith** [comment]
> Do you have plans to add a frontend feature making use of the alliance or is it just to dump the data?

Currently working on a Discord Bot that needs that data. If you think this will complicate the API unnecessarily let me know.

**Comment by kejith** [comment]
Requested changes implemented.

**Review by anoff — APPROVED** [review]

**Pull request merged**